src/java.desktop/share/classes/javax/swing/JPopupMenu.java

```diff
@@ -941,10 +941,29 @@ public Component getInvoker() {
     public void setInvoker(Component invoker) {
         Component oldInvoker = this.invoker;
         this.invoker = invoker;
+        PropertyChangeListener propListener = new PropertyChangeListener() {
+            public void propertyChange(PropertyChangeEvent e) {
+                String propertyName = e.getPropertyName();
+                if (propertyName == "ancestor") {
+                    if (e.getOldValue() != null &&
+                            e.getNewValue() == null) {
+                        if (popup != null) {
+                            popup.dispose();
+                            popup = null;
+                        }
+
+                    }
+                }
+            }
+        };
         if ((oldInvoker != this.invoker) && (ui != null)) {
             ui.uninstallUI(this);
+            if (oldInvoker != null) {
+                oldInvoker.removePropertyChangeListener(propListener);
```

Comment:
This won't remove the install `propListener` because a new instance of the listener is created each time the method is called.